<html><body><div><div style="padding-left: 10;margin-left: 25;margin-top: 40;border-left-style: solid;border-color: grey;"><h1>Olmsted Dataset</h1><p><b>Description: </b><span>Olmsted dataset input file.</span></p><p><b>Required: </b><code>['dataset_id', 'clones']</code></p><p><b>Type: </b><code>object</code></p><div><h2>Properties:</h2><div style="margin-left: 10px;"><h3><code>paper</code></h3><b style="padding-left: 15; font-size: 18;">{Paper info}</b></div><div style="margin-left: 10px;"><h3><code>ident</code></h3><div style="padding-left: 10;margin-left: 25;margin-top: 10;"><p><b>Description: </b><span>UUID specific to the given object</span></p><p><b>Type: </b><code>string</code></p></div></div><div style="margin-left: 10px;"><h3><code>build</code></h3><b style="padding-left: 15; font-size: 18;">{Build info}</b></div><div style="margin-left: 10px;"><h3><code>samples</code></h3><div style="padding-left: 10;margin-left: 25;margin-top: 10;"><p><b>Description: </b><span>Information about each of the samples</span></p><p><b>Type: </b><code>array</code></p><div><h3>Array Items:</h3><b style="padding-left: 15; font-size: 18;">{Sample}</b></div></div></div><div style="margin-left: 10px;"><h3><code>seeds</code></h3><div style="padding-left: 10;margin-left: 25;margin-top: 10;"><p><b>Description: </b><span>Information about each of the seed sequences</span></p><p><b>Type: </b><code>array</code></p><div><h3>Array Items:</h3><b style="padding-left: 15; font-size: 18;">{Seed}</b></div></div></div><div style="margin-left: 10px;"><h3><code>clones</code></h3><div style="padding-left: 10;margin-left: 25;margin-top: 10;"><p><b>Description: </b><span>Information about each of the clonal families</span></p><p><b>Type: </b><code>array</code></p><div><h3>Array Items:</h3><b style="padding-left: 15; font-size: 18;">{Clone}</b></div></div></div><div style="margin-left: 10px;"><h3><code>dataset_id</code></h3><div style="padding-left: 10;margin-left: 25;margin-top: 10;"><p><b>Description: </b><span>Unique identifier for a collection of data</span></p><p><b>Type: </b><code>string</code></p></div></div><div style="margin-left: 10px;"><h3><code>subjects</code></h3><div style="padding-left: 10;margin-left: 25;margin-top: 10;"><p><b>Description: </b><span>Information about each of the subjects</span></p><p><b>Type: </b><code>array</code></p><div><h3>Array Items:</h3><b style="padding-left: 15; font-size: 18;">{Subject}</b></div></div></div></div></div><div style="padding-left: 10;margin-left: 25;margin-top: 40;border-left-style: solid;border-color: grey;"><h1>Paper info</h1><p><b>Description: </b><span>Information about a paper corresponding to this dataset</span></p><p><b>Required: </b><code>['authorstring']</code></p><p><b>Type: </b><code>object</code></p><div><h2>Properties:</h2><div style="margin-left: 10px;"><h3><code>url</code></h3><div style="padding-left: 10;margin-left: 25;margin-top: 10;"><p><b>Description: </b><span>Link to online version of the paper.</span></p><p><b>Type: </b><code>string</code></p></div></div><div style="margin-left: 10px;"><h3><code>authorstring</code></h3><div style="padding-left: 10;margin-left: 25;margin-top: 10;"><p><b>Description: </b><span>String to be displayed citing authors, e.g. "Doe, et. al.".</span></p><p><b>Type: </b><code>string</code></p></div></div></div></div><div style="padding-left: 10;margin-left: 25;margin-top: 40;border-left-style: solid;border-color: grey;"><h1>Build info</h1><p><b>Description: </b><span>Information about how a dataset was built.</span></p><p><b>Required: </b><code>['commit']</code></p><p><b>Type: </b><code>object</code></p><div><h2>Properties:</h2><div style="margin-left: 10px;"><h3><code>commit</code></h3><div style="padding-left: 10;margin-left: 25;margin-top: 10;"><p><b>Description: </b><span>Commit sha of whatever build system you used to process the data</span></p><p><b>Type: </b><code>string</code></p></div></div><div style="margin-left: 10px;"><h3><code>time</code></h3><div style="padding-left: 10;margin-left: 25;margin-top: 10;"><p><b>Description: </b><span>Time at which build was initiated</span></p><p><b>Type: </b><code>string</code></p></div></div></div></div><div style="padding-left: 10;margin-left: 25;margin-top: 40;border-left-style: solid;border-color: grey;"><h1>Sample</h1><p><b>Description: </b><span>A sample is a collection of sequences.</span></p><p><b>Required: </b><code>['locus']</code></p><p><b>Type: </b><code>object</code></p><div><h2>Properties:</h2><div style="margin-left: 10px;"><h3><code>locus</code></h3><div style="padding-left: 10;margin-left: 25;margin-top: 10;"><p><b>Description: </b><span>B-cell Locus.</span></p><p><b>Type: </b><code>string</code></p></div></div><div style="margin-left: 10px;"><h3><code>ident</code></h3><div style="padding-left: 10;margin-left: 25;margin-top: 10;"><p><b>Description: </b><span>UUID specific to the given object</span></p><p><b>Type: </b><code>string</code></p></div></div><div style="margin-left: 10px;"><h3><code>timepoint_id</code></h3><div style="padding-left: 10;margin-left: 25;margin-top: 10;"><p><b>Description: </b><span>Timepoint associated with this sample (may choose "merged" if data has been combined from multiple timepoints)</span></p><p><b>Type: </b><code>string</code></p></div></div><div style="margin-left: 10px;"><h3><code>sample_id</code></h3><div style="padding-left: 10;margin-left: 25;margin-top: 10;"><p><b>Description: </b><span>Sample id</span></p><p><b>Type: </b><code>string</code></p></div></div></div></div><div style="padding-left: 10;margin-left: 25;margin-top: 40;border-left-style: solid;border-color: grey;"><h1>Seed</h1><p><b>Description: </b><span>A sequence of interest among other clonal family members.</span></p><p><b>Required: </b><code>['seed_id']</code></p><p><b>Type: </b><code>['object', 'null']</code></p><div><h2>Properties:</h2><div style="margin-left: 10px;"><h3><code>ident</code></h3><div style="padding-left: 10;margin-left: 25;margin-top: 10;"><p><b>Description: </b><span>UUID specific to the given object</span></p><p><b>Type: </b><code>string</code></p></div></div><div style="margin-left: 10px;"><h3><code>seed_id</code></h3><div style="padding-left: 10;margin-left: 25;margin-top: 10;"><p><b>Description: </b><span>Seed id</span></p><p><b>Type: </b><code>string</code></p></div></div></div></div><div style="padding-left: 10;margin-left: 25;margin-top: 40;border-left-style: solid;border-color: grey;"><h1>Clone</h1><p><b>Description: </b><span>Clonal family of sequences deriving from a particular reassortment event</span></p><p><b>Required: </b><code>['unique_seqs_count', 'mean_mut_freq', 'v_alignment_start', 'v_alignment_end', 'j_alignment_start', 'j_alignment_end']</code></p><p><b>Type: </b><code>object</code></p><div><h2>Properties:</h2><div style="margin-left: 10px;"><h3><code>j_call</code></h3><div style="padding-left: 10;margin-left: 25;margin-top: 10;"><p><b>Description: </b><span>AIRR: J gene with allele of the inferred ancestor of the clone. For example, IGHJ4*02.</span></p><p><b>Type: </b><code>string</code></p></div></div><div style="margin-left: 10px;"><h3><code>clone_id</code></h3><div style="padding-left: 10;margin-left: 25;margin-top: 10;"><p><b>Description: </b><span>AIRR: Identifier for the clone.</span></p><p><b>Type: </b><code>string</code></p></div></div><div style="margin-left: 10px;"><h3><code>seed_id</code></h3><div style="padding-left: 10;margin-left: 25;margin-top: 10;"><p><b>Description: </b><span>Seed sequence id if any.</span></p><p><b>Type: </b><code>['string', 'null']</code></p></div></div><div style="margin-left: 10px;"><h3><code>sample_id</code></h3><div style="padding-left: 10;margin-left: 25;margin-top: 10;"><p><b>Description: </b><span>sample id associated with this clonal family.</span></p><p><b>Type: </b><code>string</code></p></div></div><div style="margin-left: 10px;"><h3><code>v_call</code></h3><div style="padding-left: 10;margin-left: 25;margin-top: 10;"><p><b>Description: </b><span>AIRR: V gene with allele of the inferred ancestral of the clone. For example, IGHV4-59*01.</span></p><p><b>Type: </b><code>string</code></p></div></div><div style="margin-left: 10px;"><h3><code>d_call</code></h3><div style="padding-left: 10;margin-left: 25;margin-top: 10;"><p><b>Description: </b><span>AIRR: D gene with allele of the inferred ancestor of the clone. For example, IGHD3-10*01.</span></p><p><b>Type: </b><code>string</code></p></div></div><div style="margin-left: 10px;"><h3><code>subject_id</code></h3><div style="padding-left: 10;margin-left: 25;margin-top: 10;"><p><b>Description: </b><span>Id of subject from which the clonal family was sampled.</span></p><p><b>Type: </b><code>string</code></p></div></div><div style="margin-left: 10px;"><h3><code>junction_length</code></h3><div style="padding-left: 10;margin-left: 25;margin-top: 10;"><p><b>Description: </b><span>AIRR: Number of nucleotides in the junction. (see AIRR 'junction': Nucleotide sequence for the junction region of the inferred ancestor of the clone, where the junction is defined as the CDR3 plus the two flanking conserved codons.)</span></p><p><b>Type: </b><code>integer</code></p></div></div><div style="margin-left: 10px;"><h3><code>d_alignment_start</code></h3><div style="padding-left: 10;margin-left: 25;margin-top: 10;"><p><b>Description: </b><span>AIRR: Start position of the D segment in both the sequence_alignment and germline_alignment fields (1-based closed interval).</span></p><p><b>Type: </b><code>integer</code></p></div></div><div style="margin-left: 10px;"><h3><code>unique_seqs_count</code></h3><div style="padding-left: 10;margin-left: 25;margin-top: 10;"><p><b>Description: </b><span>Number of unique sequences in the clone</span></p><p><b>Type: </b><code>integer</code></p></div></div><div style="margin-left: 10px;"><h3><code>d_alignment_end</code></h3><div style="padding-left: 10;margin-left: 25;margin-top: 10;"><p><b>Description: </b><span>AIRR: End position of the D segment in both the sequence_alignment and germline_alignment fields (1-based closed interval).</span></p><p><b>Type: </b><code>integer</code></p></div></div><div style="margin-left: 10px;"><h3><code>total_read_count</code></h3><div style="padding-left: 10;margin-left: 25;margin-top: 10;"><p><b>Description: </b><span>Number of total reads represented by sequences in the clone.</span></p><p><b>Type: </b><code>integer</code></p></div></div><div style="margin-left: 10px;"><h3><code>v_alignment_start</code></h3><div style="padding-left: 10;margin-left: 25;margin-top: 10;"><p><b>Description: </b><span>AIRR: Start position in the V segment in both the sequence_alignment and germline_alignment fields (1-based closed interval).</span></p><p><b>Type: </b><code>integer</code></p></div></div><div style="margin-left: 10px;"><h3><code>trees</code></h3><div style="padding-left: 10;margin-left: 25;margin-top: 10;"><p><b>Description: </b><span>Phylogenetic trees, and possibly ancestral sequence reconstructions.</span></p><p><b>Type: </b><code>array</code></p><div><h3>Array Items:</h3><b style="padding-left: 15; font-size: 18;">{Tree}</b></div></div></div><div style="margin-left: 10px;"><h3><code>j_alignment_end</code></h3><div style="padding-left: 10;margin-left: 25;margin-top: 10;"><p><b>Description: </b><span>AIRR: End position of the J segment in both the sequence_alignment and germline_alignment fields (1-based closed interval).</span></p><p><b>Type: </b><code>integer</code></p></div></div><div style="margin-left: 10px;"><h3><code>j_alignment_start</code></h3><div style="padding-left: 10;margin-left: 25;margin-top: 10;"><p><b>Description: </b><span>AIRR: Start position of the J segment in both the sequence_alignment and germline_alignment fields (1-based closed interval).</span></p><p><b>Type: </b><code>integer</code></p></div></div><div style="margin-left: 10px;"><h3><code>has_seed</code></h3><div style="padding-left: 10;margin-left: 25;margin-top: 10;"><p><b>Description: </b><span>Does this clone have a seed sequence (see Seed schema) in it?</span></p><p><b>Type: </b><code>boolean</code></p></div></div><div style="margin-left: 10px;"><h3><code>ident</code></h3><div style="padding-left: 10;margin-left: 25;margin-top: 10;"><p><b>Description: </b><span>UUID specific to the given object</span></p><p><b>Type: </b><code>string</code></p></div></div><div style="margin-left: 10px;"><h3><code>mean_mut_freq</code></h3><div style="padding-left: 10;margin-left: 25;margin-top: 10;"><p><b>Description: </b><span>Mean mutation frequency across sequences in the clone.</span></p><p><b>Type: </b><code>number</code></p></div></div><div style="margin-left: 10px;"><h3><code>germline_alignment</code></h3><div style="padding-left: 10;margin-left: 25;margin-top: 10;"><p><b>Description: </b><span>AIRR: Assembled, aligned, full-length inferred ancestor of the clone spanning the same region as the sequence_alignment field of nodes (typically the V(D)J region) and including the same set of corrections and spacers (if any).</span></p><p><b>Type: </b><code>string</code></p></div></div><div style="margin-left: 10px;"><h3><code>v_alignment_end</code></h3><div style="padding-left: 10;margin-left: 25;margin-top: 10;"><p><b>Description: </b><span>AIRR: End position in the V segment in both the sequence_alignment and germline_alignment fields (1-based closed interval).</span></p><p><b>Type: </b><code>integer</code></p></div></div><div style="margin-left: 10px;"><h3><code>junction_start</code></h3><div style="padding-left: 10;margin-left: 25;margin-top: 10;"><p><b>Description: </b><span>AIRR: Junction region start position in the alignment (1-based closed interval).</span></p><p><b>Type: </b><code>integer</code></p></div></div></div></div><div style="padding-left: 10;margin-left: 25;margin-top: 40;border-left-style: solid;border-color: grey;"><h1>Tree</h1><p><b>Description: </b><span>Phylogenetic tree and possibly ancestral state reconstruction of sequences in a clonal family.</span></p><p><b>Required: </b><code>['newick', 'nodes']</code></p><p><b>Type: </b><code>object</code></p><div><h2>Properties:</h2><div style="margin-left: 10px;"><h3><code>ident</code></h3><div style="padding-left: 10;margin-left: 25;margin-top: 10;"><p><b>Description: </b><span>UUID specific to the given object</span></p><p><b>Type: </b><code>string</code></p></div></div><div style="margin-left: 10px;"><h3><code>newick</code></h3><div style="padding-left: 10;margin-left: 25;margin-top: 10;"><p><b>Description: </b><span>AIRR: Newick string of the tree edges.</span></p><p><b>Type: </b><code>string</code></p></div></div><div style="margin-left: 10px;"><h3><code>clone_id</code></h3><div style="padding-left: 10;margin-left: 25;margin-top: 10;"><p><b>Description: </b><span>AIRR: Identifier for the clone.</span></p><p><b>Type: </b><code>string</code></p></div></div><div style="margin-left: 10px;"><h3><code>downsampling_strategy</code></h3><div style="padding-left: 10;margin-left: 25;margin-top: 10;"><p><b>Description: </b><span>If applicable, the downsampling method applied to the set of clonal sequences before passing them to a phylogenetic inference tool.</span></p><p><b>Type: </b><code>string</code></p></div></div><div style="margin-left: 10px;"><h3><code>tree_id</code></h3><div style="padding-left: 10;margin-left: 25;margin-top: 10;"><p><b>Description: </b><span>AIRR: Identifier for the tree.</span></p><p><b>Type: </b><code>string</code></p></div></div><div style="margin-left: 10px;"><h3><code>nodes</code></h3><div style="padding-left: 10;margin-left: 25;margin-top: 40;border-left-style: solid;border-color: grey;"><p><b>Description: </b><span>AIRR: Dictionary of nodes in the tree, keyed by sequence_id string.</span></p><p><b>Type: </b><code>object</code></p><div><h3>Object with values of type:</h3><b style="padding-left: 15; font-size: 18;">{Node}</b></div></div></div><div style="margin-left: 10px;"><h3><code>downsampled_count</code></h3><div style="padding-left: 10;margin-left: 25;margin-top: 10;"><p><b>Description: </b><span>If applicable, the maximum number of sequences kept in the downsampling process.</span></p><p><b>Type: </b><code>integer</code></p></div></div></div></div><div style="padding-left: 10;margin-left: 25;margin-top: 40;border-left-style: solid;border-color: grey;"><h1>Node</h1><p><b>Description: </b><span>Information about the phylogenetic tree nodes and the sequences they represent</span></p><p><b>Required: </b><code>['sequence_id', 'sequence_alignment', 'sequence_alignment_aa']</code></p><p><b>Type: </b><code>object</code></p><div><h2>Properties:</h2><div style="margin-left: 10px;"><h3><code>sequence_alignment_aa</code></h3><div style="padding-left: 10;margin-left: 25;margin-top: 10;"><p><b>Description: </b><span>Amino acid sequence of the node, aligned to the germline_alignment for this clone, including any indel corrections or spacers.</span></p><p><b>Type: </b><code>string</code></p></div></div><div style="margin-left: 10px;"><h3><code>cluster_multiplicity</code></h3><div style="padding-left: 10;margin-left: 25;margin-top: 10;"><p><b>Description: </b><span>If clonal family sequences were downsampled by clustering, the cummulative number of times sequences in cluster were observed.</span></p><p><b>Type: </b><code>['integer', 'null']</code></p></div></div><div style="margin-left: 10px;"><h3><code>timepoint_id</code></h3><div style="padding-left: 10;margin-left: 25;margin-top: 10;"><p><b>Description: </b><span>Timepoint associated with sequence, if any.</span></p><p><b>Type: </b><code>['string', 'null']</code></p></div></div><div style="margin-left: 10px;"><h3><code>sequence_id</code></h3><div style="padding-left: 10;margin-left: 25;margin-top: 10;"><p><b>Description: </b><span>AIRR: Identifier for this node that matches the id in the newick string and, where possible, the sequence_id in the source repertoire.</span></p><p><b>Type: </b><code>string</code></p></div></div><div style="margin-left: 10px;"><h3><code>lbr</code></h3><div style="padding-left: 10;margin-left: 25;margin-top: 10;"><p><b>Description: </b><span>Local branching rate (derivative of lbi; see https://arxiv.org/abs/2004.11868).</span></p><p><b>Type: </b><code>['number', 'null']</code></p></div></div><div style="margin-left: 10px;"><h3><code>lbi</code></h3><div style="padding-left: 10;margin-left: 25;margin-top: 10;"><p><b>Description: </b><span>Local branching index (see https://arxiv.org/abs/2004.11868).</span></p><p><b>Type: </b><code>['number', 'null']</code></p></div></div><div style="margin-left: 10px;"><h3><code>cluster_timepoint_multiplicities</code></h3><div style="padding-left: 10;margin-left: 25;margin-top: 10;"><p><b>Description: </b><span>Sequence multiplicity, broken down by timepoint, including sequences falling in the same cluster if clustering-based downsampling was performed.</span></p><p><b>Type: </b><code>array</code></p><div><h3>Array Items:</h3><b style="padding-left: 15; font-size: 18;">{Timepoint multiplicity}</b></div></div></div><div style="margin-left: 10px;"><h3><code>timepoint_multiplicities</code></h3><div style="padding-left: 10;margin-left: 25;margin-top: 10;"><p><b>Description: </b><span>Sequence multiplicity, broken down by timepoint.</span></p><p><b>Type: </b><code>array</code></p><div><h3>Array Items:</h3><b style="padding-left: 15; font-size: 18;">{Timepoint multiplicity}</b></div></div></div><div style="margin-left: 10px;"><h3><code>multiplicity</code></h3><div style="padding-left: 10;margin-left: 25;margin-top: 10;"><p><b>Description: </b><span>Number of times sequence was observed in the sample. The presence of a given sequence in a clonal family may represent many identical such sequences in the original sample.</span></p><p><b>Type: </b><code>['integer', 'null']</code></p></div></div><div style="margin-left: 10px;"><h3><code>sequence_alignment</code></h3><div style="padding-left: 10;margin-left: 25;margin-top: 10;"><p><b>Description: </b><span>AIRR: Nucleotide sequence of the node, aligned to the germline_alignment for this clone, including any indel corrections or spacers.</span></p><p><b>Type: </b><code>string</code></p></div></div><div style="margin-left: 10px;"><h3><code>affinity</code></h3><div style="padding-left: 10;margin-left: 25;margin-top: 10;"><p><b>Description: </b><span>Affinity of the antibody for some antigen. Typically inverse dissociation constant k_d in simulation, and inverse ic50 in data.</span></p><p><b>Type: </b><code>['number', 'null']</code></p></div></div></div></div><div style="padding-left: 10;margin-left: 25;margin-top: 40;border-left-style: solid;border-color: grey;"><h1>Timepoint multiplicity</h1><p><b>Description: </b><span>Multiplicity at a specific time.</span></p><p><b>Type: </b><code>object</code></p><div><h2>Properties:</h2><div style="margin-left: 10px;"><h3><code>multiplicity</code></h3><div style="padding-left: 10;margin-left: 25;margin-top: 10;"><p><b>Description: </b><span>Number of times sequence was observed at the given timepoint</span></p><p><b>Type: </b><code>['integer', 'null']</code></p></div></div><div style="margin-left: 10px;"><h3><code>timepoint_id</code></h3><div style="padding-left: 10;margin-left: 25;margin-top: 10;"><p><b>Description: </b><span>Id associated with the timepoint in question</span></p><p><b>Type: </b><code>string</code></p></div></div></div></div><div style="padding-left: 10;margin-left: 25;margin-top: 40;border-left-style: solid;border-color: grey;"><h1>Subject</h1><p><b>Description: </b><span>Subject from which the clonal family was sampled.</span></p><p><b>Required: </b><code>['subject_id']</code></p><p><b>Type: </b><code>object</code></p><div><h2>Properties:</h2><div style="margin-left: 10px;"><h3><code>subject_id</code></h3><div style="padding-left: 10;margin-left: 25;margin-top: 10;"><p><b>Description: </b><span>Subject id</span></p><p><b>Type: </b><code>string</code></p></div></div><div style="margin-left: 10px;"><h3><code>ident</code></h3><div style="padding-left: 10;margin-left: 25;margin-top: 10;"><p><b>Description: </b><span>UUID specific to the given object</span></p><p><b>Type: </b><code>string</code></p></div></div></div></div></div></body></html>