<html><body><div><div style="margin-top:40; border-left-style:solid; margin-left:25; border-color:grey; padding-left:10;"><h1>Olmsted Dataset</h1><p><b>Description: </b>Olmsted dataset input file.</p><p><b>Required: </b><code>['dataset_id', 'clones']</code></p><p><b>Type: </b><code>object</code></p><div><h2>Properties:</h2><div style="margin-left:10px;"><h3><code>ident</code></h3><div style="margin-top:10; padding-left:10; margin-left:25;"><p><b>Description: </b>UUID specific to the given object</p><p><b>Type: </b><code>string</code></p></div></div><div style="margin-left:10px;"><h3><code>subjects</code></h3><div style="margin-top:10; padding-left:10; margin-left:25;"><p><b>Description: </b>Information about each of the subjects</p><p><b>Type: </b><code>array</code></p><div><h3>Array Items:</h3><b style="font-size:18; padding-left:15;">{Subject}</b></div></div></div><div style="margin-left:10px;"><h3><code>build</code></h3><b style="font-size:18; padding-left:15;">{Build info}</b></div><div style="margin-left:10px;"><h3><code>samples</code></h3><div style="margin-top:10; padding-left:10; margin-left:25;"><p><b>Description: </b>Information about each of the samples</p><p><b>Type: </b><code>array</code></p><div><h3>Array Items:</h3><b style="font-size:18; padding-left:15;">{Sample}</b></div></div></div><div style="margin-left:10px;"><h3><code>seeds</code></h3><div style="margin-top:10; padding-left:10; margin-left:25;"><p><b>Description: </b>Information about each of the seed sequences</p><p><b>Type: </b><code>array</code></p><div><h3>Array Items:</h3><b style="font-size:18; padding-left:15;">{Seed}</b></div></div></div><div style="margin-left:10px;"><h3><code>clones</code></h3><div style="margin-top:10; padding-left:10; margin-left:25;"><p><b>Description: </b>Information about each of the clonal lineages</p><p><b>Type: </b><code>array</code></p><div><h3>Array Items:</h3><b style="font-size:18; padding-left:15;">{Clone}</b></div></div></div><div style="margin-left:10px;"><h3><code>dataset_id</code></h3><div style="margin-top:10; padding-left:10; margin-left:25;"><p><b>Description: </b>Unique identifier for a collection of data</p><p><b>Type: </b><code>string</code></p></div></div></div></div><div style="margin-top:40; border-left-style:solid; margin-left:25; border-color:grey; padding-left:10;"><h1>Subject</h1><p><b>Description: </b>Subject from which the clonal lineage was sampled.</p><p><b>Required: </b><code>['subject_id']</code></p><p><b>Type: </b><code>object</code></p><div><h2>Properties:</h2><div style="margin-left:10px;"><h3><code>subject_id</code></h3><div style="margin-top:10; padding-left:10; margin-left:25;"><p><b>Description: </b>Subject id</p><p><b>Type: </b><code>string</code></p></div></div><div style="margin-left:10px;"><h3><code>ident</code></h3><div style="margin-top:10; padding-left:10; margin-left:25;"><p><b>Description: </b>UUID specific to the given object</p><p><b>Type: </b><code>string</code></p></div></div></div></div><div style="margin-top:40; border-left-style:solid; margin-left:25; border-color:grey; padding-left:10;"><h1>Build info</h1><p><b>Description: </b>Information about how a dataset was built.</p><p><b>Required: </b><code>['commit']</code></p><p><b>Type: </b><code>object</code></p><div><h2>Properties:</h2><div style="margin-left:10px;"><h3><code>commit</code></h3><div style="margin-top:10; padding-left:10; margin-left:25;"><p><b>Description: </b>Commit sha of whatever build system you used to process the data</p><p><b>Type: </b><code>string</code></p></div></div><div style="margin-left:10px;"><h3><code>time</code></h3><div style="margin-top:10; padding-left:10; margin-left:25;"><p><b>Description: </b>Time at which build was initiated</p><p><b>Type: </b><code>string</code></p></div></div></div></div><div style="margin-top:40; border-left-style:solid; margin-left:25; border-color:grey; padding-left:10;"><h1>Sample</h1><p><b>Description: </b>A sample is generally a collection of sequences.</p><p><b>Required: </b><code>['locus']</code></p><p><b>Type: </b><code>object</code></p><div><h2>Properties:</h2><div style="margin-left:10px;"><h3><code>locus</code></h3><div style="margin-top:10; padding-left:10; margin-left:25;"><p><b>Description: </b>B-cell Locus.</p><p><b>Type: </b><code>string</code></p></div></div><div style="margin-left:10px;"><h3><code>ident</code></h3><div style="margin-top:10; padding-left:10; margin-left:25;"><p><b>Description: </b>UUID specific to the given object</p><p><b>Type: </b><code>string</code></p></div></div><div style="margin-left:10px;"><h3><code>timepoint_id</code></h3><div style="margin-top:10; padding-left:10; margin-left:25;"><p><b>Description: </b>Timepoint associated with this sample (may choose "merged" if data has been combined from multiple timepoints)</p><p><b>Type: </b><code>string</code></p></div></div><div style="margin-left:10px;"><h3><code>sample_id</code></h3><div style="margin-top:10; padding-left:10; margin-left:25;"><p><b>Description: </b>Sample id</p><p><b>Type: </b><code>string</code></p></div></div></div></div><div style="margin-top:40; border-left-style:solid; margin-left:25; border-color:grey; padding-left:10;"><h1>Seed</h1><p><b>Description: </b>A sequence of interest among other clonal lineage members.</p><p><b>Required: </b><code>['seed_id']</code></p><p><b>Type: </b><code>['object', 'null']</code></p><div><h2>Properties:</h2><div style="margin-left:10px;"><h3><code>ident</code></h3><div style="margin-top:10; padding-left:10; margin-left:25;"><p><b>Description: </b>UUID specific to the given object</p><p><b>Type: </b><code>string</code></p></div></div><div style="margin-left:10px;"><h3><code>seed_id</code></h3><div style="margin-top:10; padding-left:10; margin-left:25;"><p><b>Description: </b>Seed id</p><p><b>Type: </b><code>string</code></p></div></div></div></div><div style="margin-top:40; border-left-style:solid; margin-left:25; border-color:grey; padding-left:10;"><h1>Clone</h1><p><b>Description: </b>Clonal lineage of sequences deriving from a particular reassortment event</p><p><b>Required: </b><code>['unique_seqs_count', 'mean_mut_freq', 'v_alignment_start', 'v_alignment_end', 'j_alignment_start', 'j_alignment_end']</code></p><p><b>Type: </b><code>object</code></p><div><h2>Properties:</h2><div style="margin-left:10px;"><h3><code>j_call</code></h3><div style="margin-top:10; padding-left:10; margin-left:25;"><p><b>Description: </b>AIRR: J gene with allele of the inferred ancestor of the clone. For example, IGHJ4*02.</p><p><b>Type: </b><code>string</code></p></div></div><div style="margin-left:10px;"><h3><code>clone_id</code></h3><div style="margin-top:10; padding-left:10; margin-left:25;"><p><b>Description: </b>AIRR: Identifier for the clone.</p><p><b>Type: </b><code>string</code></p></div></div><div style="margin-left:10px;"><h3><code>seed_id</code></h3><div style="margin-top:10; padding-left:10; margin-left:25;"><p><b>Description: </b>Seed sequence id if any.</p><p><b>Type: </b><code>['string', 'null']</code></p></div></div><div style="margin-left:10px;"><h3><code>sample_id</code></h3><div style="margin-top:10; padding-left:10; margin-left:25;"><p><b>Description: </b>sample id associated with this clonal lineage.</p><p><b>Type: </b><code>string</code></p></div></div><div style="margin-left:10px;"><h3><code>v_call</code></h3><div style="margin-top:10; padding-left:10; margin-left:25;"><p><b>Description: </b>AIRR: V gene with allele of the inferred ancestral of the clone. For example, IGHV4-59*01.</p><p><b>Type: </b><code>string</code></p></div></div><div style="margin-left:10px;"><h3><code>d_call</code></h3><div style="margin-top:10; padding-left:10; margin-left:25;"><p><b>Description: </b>AIRR: D gene with allele of the inferred ancestor of the clone. For example, IGHD3-10*01.</p><p><b>Type: </b><code>string</code></p></div></div><div style="margin-left:10px;"><h3><code>subject_id</code></h3><div style="margin-top:10; padding-left:10; margin-left:25;"><p><b>Description: </b>Id of subject from which the clonal lineage was sampled.</p><p><b>Type: </b><code>string</code></p></div></div><div style="margin-left:10px;"><h3><code>junction_length</code></h3><div style="margin-top:10; padding-left:10; margin-left:25;"><p><b>Description: </b>AIRR: Number of nucleotides in the junction. (see AIRR 'junction': Nucleotide sequence for the junction region of the inferred ancestor of the clone, where the junction is defined as the CDR3 plus the two flanking conserved codons.)</p><p><b>Type: </b><code>integer</code></p></div></div><div style="margin-left:10px;"><h3><code>d_alignment_start</code></h3><div style="margin-top:10; padding-left:10; margin-left:25;"><p><b>Description: </b>AIRR: Start position of the D segment in both the sequence_alignment and germline_alignment fields (1-based closed interval).</p><p><b>Type: </b><code>integer</code></p></div></div><div style="margin-left:10px;"><h3><code>unique_seqs_count</code></h3><div style="margin-top:10; padding-left:10; margin-left:25;"><p><b>Description: </b>Number of unique sequences in the clone</p><p><b>Type: </b><code>integer</code></p></div></div><div style="margin-left:10px;"><h3><code>d_alignment_end</code></h3><div style="margin-top:10; padding-left:10; margin-left:25;"><p><b>Description: </b>AIRR: End position of the D segment in both the sequence_alignment and germline_alignment fields (1-based closed interval).</p><p><b>Type: </b><code>integer</code></p></div></div><div style="margin-left:10px;"><h3><code>total_read_count</code></h3><div style="margin-top:10; padding-left:10; margin-left:25;"><p><b>Description: </b>Number of total reads represented by sequences in the clone.</p><p><b>Type: </b><code>integer</code></p></div></div><div style="margin-left:10px;"><h3><code>v_alignment_start</code></h3><div style="margin-top:10; padding-left:10; margin-left:25;"><p><b>Description: </b>AIRR: Start position in the V segment in both the sequence_alignment and germline_alignment fields (1-based closed interval).</p><p><b>Type: </b><code>integer</code></p></div></div><div style="margin-left:10px;"><h3><code>trees</code></h3><div style="margin-top:10; padding-left:10; margin-left:25;"><p><b>Description: </b>Phylogenetic trees, and possibly ancestral sequence reconstructions.</p><p><b>Type: </b><code>array</code></p><div><h3>Array Items:</h3><b style="font-size:18; padding-left:15;">{Tree}</b></div></div></div><div style="margin-left:10px;"><h3><code>j_alignment_end</code></h3><div style="margin-top:10; padding-left:10; margin-left:25;"><p><b>Description: </b>AIRR: End position of the J segment in both the sequence_alignment and germline_alignment fields (1-based closed interval).</p><p><b>Type: </b><code>integer</code></p></div></div><div style="margin-left:10px;"><h3><code>j_alignment_start</code></h3><div style="margin-top:10; padding-left:10; margin-left:25;"><p><b>Description: </b>AIRR: Start position of the J segment in both the sequence_alignment and germline_alignment fields (1-based closed interval).</p><p><b>Type: </b><code>integer</code></p></div></div><div style="margin-left:10px;"><h3><code>has_seed</code></h3><div style="margin-top:10; padding-left:10; margin-left:25;"><p><b>Description: </b>Does this clone have a seed sequence in it?</p><p><b>Type: </b><code>boolean</code></p></div></div><div style="margin-left:10px;"><h3><code>ident</code></h3><div style="margin-top:10; padding-left:10; margin-left:25;"><p><b>Description: </b>UUID specific to the given object</p><p><b>Type: </b><code>string</code></p></div></div><div style="margin-left:10px;"><h3><code>mean_mut_freq</code></h3><div style="margin-top:10; padding-left:10; margin-left:25;"><p><b>Description: </b>Mean mutation frequency across sequences in the clone.</p><p><b>Type: </b><code>number</code></p></div></div><div style="margin-left:10px;"><h3><code>germline_alignment</code></h3><div style="margin-top:10; padding-left:10; margin-left:25;"><p><b>Description: </b>AIRR: Assembled, aligned, full-length inferred ancestor of the clone spanning the same region as the sequence_alignment field of nodes (typically the V(D)J region) and including the same set of corrections and spacers (if any).</p><p><b>Type: </b><code>string</code></p></div></div><div style="margin-left:10px;"><h3><code>v_alignment_end</code></h3><div style="margin-top:10; padding-left:10; margin-left:25;"><p><b>Description: </b>AIRR: End position in the V segment in both the sequence_alignment and germline_alignment fields (1-based closed interval).</p><p><b>Type: </b><code>integer</code></p></div></div><div style="margin-left:10px;"><h3><code>junction_start</code></h3><div style="margin-top:10; padding-left:10; margin-left:25;"><p><b>Description: </b>AIRR: Junction region start position in the alignment (1-based closed interval).</p><p><b>Type: </b><code>integer</code></p></div></div></div></div><div style="margin-top:40; border-left-style:solid; margin-left:25; border-color:grey; padding-left:10;"><h1>Tree</h1><p><b>Description: </b>Phylogenetic tree and possibly ancestral state reconstruction of sequences in a clonal lineage.</p><p><b>Required: </b><code>['newick', 'nodes']</code></p><p><b>Type: </b><code>object</code></p><div><h2>Properties:</h2><div style="margin-left:10px;"><h3><code>ident</code></h3><div style="margin-top:10; padding-left:10; margin-left:25;"><p><b>Description: </b>UUID specific to the given object</p><p><b>Type: </b><code>string</code></p></div></div><div style="margin-left:10px;"><h3><code>newick</code></h3><div style="margin-top:10; padding-left:10; margin-left:25;"><p><b>Description: </b>AIRR: Newick string of the tree edges.</p><p><b>Type: </b><code>string</code></p></div></div><div style="margin-left:10px;"><h3><code>clone_id</code></h3><div style="margin-top:10; padding-left:10; margin-left:25;"><p><b>Description: </b>AIRR: Identifier for the clone.</p><p><b>Type: </b><code>string</code></p></div></div><div style="margin-left:10px;"><h3><code>downsampling_strategy</code></h3><div style="margin-top:10; padding-left:10; margin-left:25;"><p><b>Description: </b>If applicable, the downsampling method</p><p><b>Type: </b><code>string</code></p></div></div><div style="margin-left:10px;"><h3><code>tree_id</code></h3><div style="margin-top:10; padding-left:10; margin-left:25;"><p><b>Description: </b>AIRR: Identifier for the tree.</p><p><b>Type: </b><code>string</code></p></div></div><div style="margin-left:10px;"><h3><code>nodes</code></h3><div style="margin-top:40; border-left-style:solid; margin-left:25; border-color:grey; padding-left:10;"><p><b>Description: </b>AIRR: Dictionary of nodes in the tree, keyed by sequence_id string.</p><p><b>Type: </b><code>object</code></p></div></div><div style="margin-left:10px;"><h3><code>downsampled_count</code></h3><div style="margin-top:10; padding-left:10; margin-left:25;"><p><b>Description: </b>If applicable, the maximum number of sequences kept in the downsampling process</p><p><b>Type: </b><code>integer</code></p></div></div></div></div></div></body></html>