<html><body><div><div style="margin-top:40; border-left-style:solid; margin-left:25; border-color:grey; padding-left:10;"><h1>Olmsted Dataset</h1><p><b>Description: </b>Olmsted dataset input file. AIRR: see Study</p><p><b>Required: </b><code>['id', 'clonal_families']</code></p><p><b>Type: </b><code>object</code></p><div><h2>Properties:</h2><div style="margin-left:10px;"><h3><code>ident</code></h3><div style="margin-top:10; padding-left:10; margin-left:25;"><p><b>Description: </b>UUID specific to the given object</p><p><b>Type: </b><code>string</code></p></div></div><div style="margin-left:10px;"><h3><code>clonal_families</code></h3><div style="margin-top:10; padding-left:10; margin-left:25;"><p><b>Description: </b>Information about each of the clonal families</p><p><b>Type: </b><code>array</code></p><div><h3>Array Items:</h3><b style="font-size:18; padding-left:15;">{Clonal Family}</b></div></div></div><div style="margin-left:10px;"><h3><code>id</code></h3><div style="margin-top:10; padding-left:10; margin-left:25;"><p><b>Description: </b>Unique identifier for a collection of data</p><p><b>Type: </b><code>string</code></p></div></div><div style="margin-left:10px;"><h3><code>seeds</code></h3><div style="margin-top:10; padding-left:10; margin-left:25;"><p><b>Description: </b>Information about each of the seed sequences</p><p><b>Type: </b><code>array</code></p><div><h3>Array Items:</h3><b style="font-size:18; padding-left:15;">{Seed}</b></div></div></div><div style="margin-left:10px;"><h3><code>build</code></h3><b style="font-size:18; padding-left:15;">{Build info}</b></div><div style="margin-left:10px;"><h3><code>samples</code></h3><div style="margin-top:10; padding-left:10; margin-left:25;"><p><b>Description: </b>Information about each of the samples</p><p><b>Type: </b><code>array</code></p><div><h3>Array Items:</h3><b style="font-size:18; padding-left:15;">{Sample}</b></div></div></div><div style="margin-left:10px;"><h3><code>subjects</code></h3><div style="margin-top:10; padding-left:10; margin-left:25;"><p><b>Description: </b>Information about each of the subjects</p><p><b>Type: </b><code>array</code></p><div><h3>Array Items:</h3><b style="font-size:18; padding-left:15;">{Subject}</b></div></div></div></div></div><div style="margin-top:40; border-left-style:solid; margin-left:25; border-color:grey; padding-left:10;"><h1>Clonal Family</h1><p><b>Description: </b>Clonal family of sequences deriving from a particular reassortment event</p><p><b>Required: </b><code>['unique_seqs_count', 'mean_mut_freq', 'v_start', 'v_end', 'j_start', 'j_end']</code></p><p><b>Type: </b><code>object</code></p><div><h2>Properties:</h2><div style="margin-left:10px;"><h3><code>j_end</code></h3><div style="margin-top:10; padding-left:10; margin-left:25;"><p><b>Description: </b>Position in j gene at which rearrangement ends. AIRR: see j_germline_end, which uses 1-based closed interval as opposed to 0-based python slice convention intervals used by partis.</p><p><b>Type: </b><code>integer</code></p></div></div><div style="margin-left:10px;"><h3><code>seed_id</code></h3><div style="margin-top:10; padding-left:10; margin-left:25;"><p><b>Description: </b>Seed sequence id if any. AIRR: ?</p><p><b>Type: </b><code>['string', 'null']</code></p></div></div><div style="margin-left:10px;"><h3><code>sample_id</code></h3><div style="margin-top:10; padding-left:10; margin-left:25;"><p><b>Description: </b>sample id associated with this clonal family. AIRR: see sample_id</p><p><b>Type: </b><code>string</code></p></div></div><div style="margin-left:10px;"><h3><code>j_gene</code></h3><div style="margin-top:10; padding-left:10; margin-left:25;"><p><b>Description: </b>J gene used in rearrangement. AIRR: see j_call</p><p><b>Type: </b><code>string</code></p></div></div><div style="margin-left:10px;"><h3><code>id</code></h3><div style="margin-top:10; padding-left:10; margin-left:25;"><p><b>Description: </b>Clonal family id. AIRR: see rearrangement_id, clone_id</p><p><b>Type: </b><code>string</code></p></div></div><div style="margin-left:10px;"><h3><code>d_start</code></h3><div style="margin-top:10; padding-left:10; margin-left:25;"><p><b>Description: </b>Position in d gene at which rearrangement starts. AIRR: see d_germline_start, which uses 1-based closed interval as opposed to 0-based python slice convention intervals used by partis.</p><p><b>Type: </b><code>integer</code></p></div></div><div style="margin-left:10px;"><h3><code>naive_seq</code></h3><div style="margin-top:10; padding-left:10; margin-left:25;"><p><b>Description: </b>Naive nucleotide sequence. AIRR: see germline_alignment</p><p><b>Type: </b><code>string</code></p></div></div><div style="margin-left:10px;"><h3><code>cdr3_length</code></h3><div style="margin-top:10; padding-left:10; margin-left:25;"><p><b>Description: </b>Length of CDR3 region. AIRR: see junction_length</p><p><b>Type: </b><code>integer</code></p></div></div><div style="margin-left:10px;"><h3><code>subject_id</code></h3><div style="margin-top:10; padding-left:10; margin-left:25;"><p><b>Description: </b>Id of subject from which the clonal family was sampled. AIRR: see subject_id</p><p><b>Type: </b><code>string</code></p></div></div><div style="margin-left:10px;"><h3><code>j_start</code></h3><div style="margin-top:10; padding-left:10; margin-left:25;"><p><b>Description: </b>Position in j gene at which rearrangement starts. AIRR: see j_germline_start, which uses 1-based closed interval as opposed to 0-based python slice convention intervals used by partis.</p><p><b>Type: </b><code>integer</code></p></div></div><div style="margin-left:10px;"><h3><code>d_end</code></h3><div style="margin-top:10; padding-left:10; margin-left:25;"><p><b>Description: </b>Position in d gene at which rearrangement ends. AIRR: see d_germline_end, which uses 1-based closed interval as opposed to 0-based python slice convention intervals used by partis.</p><p><b>Type: </b><code>integer</code></p></div></div><div style="margin-left:10px;"><h3><code>unique_seqs_count</code></h3><div style="margin-top:10; padding-left:10; margin-left:25;"><p><b>Description: </b>Number of unique sequences in the clonal family. AIRR: ?</p><p><b>Type: </b><code>integer</code></p></div></div><div style="margin-left:10px;"><h3><code>v_start</code></h3><div style="margin-top:10; padding-left:10; margin-left:25;"><p><b>Description: </b>Position in v gene at which rearrangement starts. AIRR: see v_germline_start, which uses 1-based closed interval as opposed to 0-based python slice convention intervals used by partis.</p><p><b>Type: </b><code>integer</code></p></div></div><div style="margin-left:10px;"><h3><code>total_read_count</code></h3><div style="margin-top:10; padding-left:10; margin-left:25;"><p><b>Description: </b>Number of total reads represented by sequences in the clonal family. AIRR: ?</p><p><b>Type: </b><code>integer</code></p></div></div><div style="margin-left:10px;"><h3><code>trees</code></h3><div style="margin-top:10; padding-left:10; margin-left:25;"><p><b>Description: </b>Phylogenetic trees, and possibly ancestral sequence reconstructions. AIRR: ?</p><p><b>Type: </b><code>array</code></p><div><h3>Array Items:</h3><b style="font-size:18; padding-left:15;">{Tree}</b></div></div></div><div style="margin-left:10px;"><h3><code>v_gene</code></h3><div style="margin-top:10; padding-left:10; margin-left:25;"><p><b>Description: </b>V gene used in rearrangement. AIRR: see v_call</p><p><b>Type: </b><code>string</code></p></div></div><div style="margin-left:10px;"><h3><code>has_seed</code></h3><div style="margin-top:10; padding-left:10; margin-left:25;"><p><b>Description: </b>Does this clonal family have a seed sequence in it?. AIRR: ?</p><p><b>Type: </b><code>boolean</code></p></div></div><div style="margin-left:10px;"><h3><code>ident</code></h3><div style="margin-top:10; padding-left:10; margin-left:25;"><p><b>Description: </b>UUID specific to the given object</p><p><b>Type: </b><code>string</code></p></div></div><div style="margin-left:10px;"><h3><code>v_end</code></h3><div style="margin-top:10; padding-left:10; margin-left:25;"><p><b>Description: </b>Position in v gene at which rearrangement ends. AIRR: see v_germline_end, which uses 1-based closed interval as opposed to 0-based python slice convention intervals used by partis.</p><p><b>Type: </b><code>integer</code></p></div></div><div style="margin-left:10px;"><h3><code>mean_mut_freq</code></h3><div style="margin-top:10; padding-left:10; margin-left:25;"><p><b>Description: </b>Mean mutation frequency across sequences in the clonal family. AIRR: ?</p><p><b>Type: </b><code>number</code></p></div></div><div style="margin-left:10px;"><h3><code>d_gene</code></h3><div style="margin-top:10; padding-left:10; margin-left:25;"><p><b>Description: </b>D gene used in rearrangement. AIRR: see d_call</p><p><b>Type: </b><code>string</code></p></div></div><div style="margin-left:10px;"><h3><code>cdr3_start</code></h3><div style="margin-top:10; padding-left:10; margin-left:25;"><p><b>Description: </b>Start of the CDR3 region. From partis 'zero-indexed indel-reversed-sequence positions of the conserved cyst and tryp/phen codons'. AIRR: see cdr3_start, which excludes the conserved residue and uses a '1-based closed interval'</p><p><b>Type: </b><code>integer</code></p></div></div></div></div><div style="margin-top:40; border-left-style:solid; margin-left:25; border-color:grey; padding-left:10;"><h1>Tree</h1><p><b>Description: </b>Phylogenetic tree and possibly ancestral state reconstruction of sequences in a clonal family. AIRR: no trees in airr</p><p><b>Required: </b><code>['newick', 'nodes']</code></p><p><b>Type: </b><code>object</code></p><div><h2>Properties:</h2><div style="margin-left:10px;"><h3><code>ident</code></h3><div style="margin-top:10; padding-left:10; margin-left:25;"><p><b>Description: </b>UUID specific to the given object</p><p><b>Type: </b><code>string</code></p></div></div><div style="margin-left:10px;"><h3><code>newick</code></h3><div style="margin-top:10; padding-left:10; margin-left:25;"><p><b>Description: </b>Tree in newick format</p><p><b>Type: </b><code>string</code></p></div></div><div style="margin-left:10px;"><h3><code>downsampling_strategy</code></h3><div style="margin-top:10; padding-left:10; margin-left:25;"><p><b>Description: </b>If applicable, the downsampling method</p><p><b>Type: </b><code>string</code></p></div></div><div style="margin-left:10px;"><h3><code>nodes</code></h3><div style="margin-top:10; padding-left:10; margin-left:25;"><p><b>Description: </b>Nodes in the clonal family tree and corresponding sequences and metadata</p><p><b>Type: </b><code>array</code></p><div><h3>Array Items:</h3><b style="font-size:18; padding-left:15;">{Node}</b></div></div></div><div style="margin-left:10px;"><h3><code>id</code></h3><div style="margin-top:10; padding-left:10; margin-left:25;"><p><b>Description: </b>tree id</p><p><b>Type: </b><code>string</code></p></div></div><div style="margin-left:10px;"><h3><code>downsampled_count</code></h3><div style="margin-top:10; padding-left:10; margin-left:25;"><p><b>Description: </b>If applicable, the maximum number of sequences kept in the downsampling process</p><p><b>Type: </b><code>integer</code></p></div></div></div></div><div style="margin-top:40; border-left-style:solid; margin-left:25; border-color:grey; padding-left:10;"><h1>Node</h1><p><b>Description: </b>Information about the phylogenetic tree nodes and the sequences they represent</p><p><b>Required: </b><code>['id', 'dna_seq', 'aa_seq']</code></p><p><b>Type: </b><code>object</code></p><div><h2>Properties:</h2><div style="margin-left:10px;"><h3><code>lbr</code></h3><div style="margin-top:10; padding-left:10; margin-left:25;"><p><b>Description: </b>Local branching rate (derivative of lbi). AIRR: no trees in airr</p><p><b>Type: </b><code>['number', 'null']</code></p></div></div><div style="margin-left:10px;"><h3><code>cluster_multiplicity</code></h3><div style="margin-top:10; padding-left:10; margin-left:25;"><p><b>Description: </b>If clonal family sequences were downsampled by clustering, the cummulative number of times sequences in cluster were observed. AIRR: ?</p><p><b>Type: </b><code>['integer', 'null']</code></p></div></div><div style="margin-left:10px;"><h3><code>timepoint_id</code></h3><div style="margin-top:10; padding-left:10; margin-left:25;"><p><b>Description: </b>Timepoint associated with sequence, if any. AIRR: see collection_time_point_relative</p><p><b>Type: </b><code>['string', 'null']</code></p></div></div><div style="margin-left:10px;"><h3><code>aa_seq</code></h3><div style="margin-top:10; padding-left:10; margin-left:25;"><p><b>Description: </b>Literal amino acid sequence, aligned to other sequences in clonal family. AIRR: see sequence_aa</p><p><b>Type: </b><code>string</code></p></div></div><div style="margin-left:10px;"><h3><code>lbi</code></h3><div style="margin-top:10; padding-left:10; margin-left:25;"><p><b>Description: </b>Local branching index. AIRR: no trees in airr</p><p><b>Type: </b><code>['number', 'null']</code></p></div></div><div style="margin-left:10px;"><h3><code>id</code></h3><div style="margin-top:10; padding-left:10; margin-left:25;"><p><b>Description: </b>Sequence id. AIRR: see sequence_id</p><p><b>Type: </b><code>string</code></p></div></div><div style="margin-left:10px;"><h3><code>cluster_timepoint_multiplicities</code></h3><div style="margin-top:10; padding-left:10; margin-left:25;"><p><b>Description: </b>Sequence multiplicity, broken down by timepoint, including sequences falling in the same cluster if clustering-based downsampling was performed. AIRR: ?</p><p><b>Type: </b><code>array</code></p><div><h3>Array Items:</h3><div style="margin-top:40; border-left-style:solid; margin-left:25; border-color:grey; padding-left:10;"><p><b>Description: </b>Multiplicity at a specific time. AIRR: ?</p><p><b>Type: </b><code>object</code></p><div><h3>Properties:</h3><div style="margin-left:10px;"><h3><code>multiplicity</code></h3><div style="margin-top:10; padding-left:10; margin-left:25;"><p><b>Description: </b>Number of times sequence was observed at the given timepoint</p><p><b>Type: </b><code>['integer', 'null']</code></p></div></div><div style="margin-left:10px;"><h3><code>timepoint_id</code></h3><div style="margin-top:10; padding-left:10; margin-left:25;"><p><b>Description: </b>Id associated with the timepoint in question</p><p><b>Type: </b><code>string</code></p></div></div></div></div></div></div></div><div style="margin-left:10px;"><h3><code>timepoint_multiplicities</code></h3><div style="margin-top:10; padding-left:10; margin-left:25;"><p><b>Description: </b>Sequence multiplicity, broken down by timepoint. AIRR: ?</p><p><b>Type: </b><code>array</code></p><div><h3>Array Items:</h3><div style="margin-top:40; border-left-style:solid; margin-left:25; border-color:grey; padding-left:10;"><p><b>Description: </b>Multiplicity at a specific time. AIRR: ?</p><p><b>Type: </b><code>object</code></p><div><h3>Properties:</h3><div style="margin-left:10px;"><h3><code>multiplicity</code></h3><div style="margin-top:10; padding-left:10; margin-left:25;"><p><b>Description: </b>Number of times sequence was observed at the given timepoint</p><p><b>Type: </b><code>['integer', 'null']</code></p></div></div><div style="margin-left:10px;"><h3><code>timepoint_id</code></h3><div style="margin-top:10; padding-left:10; margin-left:25;"><p><b>Description: </b>Id associated with the timepoint in question</p><p><b>Type: </b><code>string</code></p></div></div></div></div></div></div></div><div style="margin-left:10px;"><h3><code>multiplicity</code></h3><div style="margin-top:10; padding-left:10; margin-left:25;"><p><b>Description: </b>Number of times sequence was observed in the sample. AIRR: see duplicate_count, consensus_count</p><p><b>Type: </b><code>['integer', 'null']</code></p></div></div><div style="margin-left:10px;"><h3><code>dna_seq</code></h3><div style="margin-top:10; padding-left:10; margin-left:25;"><p><b>Description: </b>Literal nucleotide sequence, aligned to other sequences in clonal family. AIRR: see sequence</p><p><b>Type: </b><code>string</code></p></div></div><div style="margin-left:10px;"><h3><code>affinity</code></h3><div style="margin-top:10; padding-left:10; margin-left:25;"><p><b>Description: </b>Affinity of the antibody for some antigen. Typically inverse dissociation constant k_d in simulation, and inverse ic50 in data. AIRR: no trees in airr</p><p><b>Type: </b><code>['number', 'null']</code></p></div></div></div></div><div style="margin-top:40; border-left-style:solid; margin-left:25; border-color:grey; padding-left:10;"><h1>Seed</h1><p><b>Description: </b>A sequence of interest among other clonal family members. AIRR: see query sequence (which is not a key but a concept around which the schema seems oriented). Unsure if all sequences in sequencing data are recorded as query sequences or just subset of interest.</p><p><b>Required: </b><code>['id']</code></p><p><b>Type: </b><code>['object', 'null']</code></p><div><h2>Properties:</h2><div style="margin-left:10px;"><h3><code>ident</code></h3><div style="margin-top:10; padding-left:10; margin-left:25;"><p><b>Description: </b>UUID specific to the given object</p><p><b>Type: </b><code>string</code></p></div></div><div style="margin-left:10px;"><h3><code>id</code></h3><div style="margin-top:10; padding-left:10; margin-left:25;"><p><b>Description: </b>Seed id</p><p><b>Type: </b><code>string</code></p></div></div></div></div><div style="margin-top:40; border-left-style:solid; margin-left:25; border-color:grey; padding-left:10;"><h1>Build info</h1><p><b>Description: </b>Information about how a dataset was built. AIRR: see DataProcessing</p><p><b>Required: </b><code>['commit']</code></p><p><b>Type: </b><code>object</code></p><div><h2>Properties:</h2><div style="margin-left:10px;"><h3><code>commit</code></h3><div style="margin-top:10; padding-left:10; margin-left:25;"><p><b>Description: </b>Commit sha of whatever build system you used to process the data</p><p><b>Type: </b><code>string</code></p></div></div><div style="margin-left:10px;"><h3><code>time</code></h3><div style="margin-top:10; padding-left:10; margin-left:25;"><p><b>Description: </b>Time at which build was initiated</p><p><b>Type: </b><code>string</code></p></div></div></div></div><div style="margin-top:40; border-left-style:solid; margin-left:25; border-color:grey; padding-left:10;"><h1>Sample</h1><p><b>Description: </b>A sample is generally a collection of sequences. AIRR: see Sample</p><p><b>Required: </b><code>['locus']</code></p><p><b>Type: </b><code>object</code></p><div><h2>Properties:</h2><div style="margin-left:10px;"><h3><code>locus</code></h3><div style="margin-top:10; padding-left:10; margin-left:25;"><p><b>Description: </b>B-cell Locus. AIRR: see locus</p><p><b>Type: </b><code>string</code></p></div></div><div style="margin-left:10px;"><h3><code>ident</code></h3><div style="margin-top:10; padding-left:10; margin-left:25;"><p><b>Description: </b>UUID specific to the given object</p><p><b>Type: </b><code>string</code></p></div></div><div style="margin-left:10px;"><h3><code>id</code></h3><div style="margin-top:10; padding-left:10; margin-left:25;"><p><b>Description: </b>Sample id</p><p><b>Type: </b><code>string</code></p></div></div><div style="margin-left:10px;"><h3><code>timepoint_id</code></h3><div style="margin-top:10; padding-left:10; margin-left:25;"><p><b>Description: </b>Timepoint associated with this sample (may choose 'merged' if data has been combined from multiple timepoints)</p><p><b>Type: </b><code>string</code></p></div></div></div></div><div style="margin-top:40; border-left-style:solid; margin-left:25; border-color:grey; padding-left:10;"><h1>Subject</h1><p><b>Description: </b>Subject from which the clonal family was sampled. AIRR: see Subject</p><p><b>Required: </b><code>['id']</code></p><p><b>Type: </b><code>object</code></p><div><h2>Properties:</h2><div style="margin-left:10px;"><h3><code>ident</code></h3><div style="margin-top:10; padding-left:10; margin-left:25;"><p><b>Description: </b>UUID specific to the given object</p><p><b>Type: </b><code>string</code></p></div></div><div style="margin-left:10px;"><h3><code>id</code></h3><div style="margin-top:10; padding-left:10; margin-left:25;"><p><b>Description: </b>Subjectd id</p><p><b>Type: </b><code>string</code></p></div></div></div></div></div></body></html>